core: Change ssa val `__repr__`

AntonLydike opened the pull request:
Change how SSA Values are printed to cli:

 - Block arguments: `BlockArgument<%0: !i32>(num_uses=0)`
 - Operation result: `OpResult<%0: !i32>(num_uses=0, op_name=arith.constant)` or `OpResult<%const: !i32>(num_uses=0, op_name=arith.constant)` if it has a `.name` attribute

PapyChacal commented on xdsl/ir.py:
```diff
@@ -204,9 +204,13 @@ def owner(self) -> Operation:
-        return f"OpResult(typ={repr(self.typ)}, num_uses={repr(len(self.uses))}, " + \
-               f"op_name={repr(self.op.name)}, " + \
-               f"result_index={repr(self.result_index)}, name={repr(self.name)})"
+        return "{}<%{}: {}>(num_uses={}, op_name={})".format(
+            self.__class__.__name__,
+            self.name or self.result_index,
```
I like the format, but I'm not sure about the self.result_index.
I'm mostly worried about cases where %{result_index} might collide with another %{self.name}, which could be confusing.
Could we use another prefix for this case ? Wondering what others think.

math-fehr replied:
Yeah I agree on this.

AntonLydike replied:
result index can never collide with a name though, right? because you can't have names end in numbers? Or am I misunderstanding something?

PapyChacal commented on xdsl/ir.py:
```diff
@@ -231,13 +235,12 @@ def owner(self) -> Block:
-               f", block={block_repr}, index={repr(self.index)}"
+        return "{}<%{}: {}>(num_uses={})".format(
+            self.__class__.__name__,
+            self.index,
```
I think Block Arguments can have names too, no? Could we try to use it here too?

math-fehr replied:
Oh yes good point, I missed that!

PapyChacal reviewed (commented):
Thanks!
I left some nitpick comments.
Generally, I'm wondering if this wouldn't be a better fit for __str__ than __repr__? I don't have a strong opinion on this, I don't know the pythonic usages in many projects.

math-fehr commented on xdsl/ir.py:
```diff
@@ -204,9 +204,13 @@ def owner(self) -> Operation:
-        return f"OpResult(typ={repr(self.typ)}, num_uses={repr(len(self.uses))}, " + \
-               f"op_name={repr(self.op.name)}, " + \
-               f"result_index={repr(self.result_index)}, name={repr(self.name)})"
+        return "{}<%{}: {}>(num_uses={}, op_name={})".format(
```
I would write it `{}(name={}, type={}, result_index={}, num_uses={}, op_name={})` instead.
Having this `<%"name": type>`  is a bit confusing, especially since names wouldn't stay the same once you show it in a large program, and also since you don't see the `result_index` when the value has a name.

math-fehr commented on xdsl/ir.py:
```diff
@@ -231,13 +235,12 @@ def owner(self) -> Block:
-            block_repr = repr(self.block)
-        return f"OpResult(typ={repr(self.typ)}, num_uses={repr(len(self.uses))}" + \
-               f", block={block_repr}, index={repr(self.index)}"
+        return "{}<%{}: {}>(num_uses={})".format(
```
Same here.

math-fehr reviewed (commented):
Nice! Thanks for taking care of this issue.

I'm just not sure we want to have the `%{}: {}` syntax, since it may be confusing for the user, especially since you won't see it's index in the operation.

math-fehr commented:
So my feeling about `str` versus `repr`, is that in Python, `repr` is supposed to always be able to be interpreted back to the object itself. However, in our case, this is simply not possible, without rewriting the entire program in `__repr__`. The reason why we redefined `repr` was that the error messages that we use to get were hundred of lines long, and since everyting is a cyclic reference anyway, this was impossible to read.

PapyChacal commented:
Oh, I get your point for `__repr__`. Makes a lot of sense then!

tobiasgrosser commented:
I am also a little bit confused about why we read a %xx here. Can we explain this in the commit message and maybe in a comment.

In MLIR/LLVM, the ->dump functions actually construct a printer to get the %xx number relative to the module correct. Without doing this, I do not see how we can print a valid number here.

superlopuh commented on xdsl/ir.py:
```diff
@@ -204,9 +204,13 @@ def owner(self) -> Operation:
-        return f"OpResult(typ={repr(self.typ)}, num_uses={repr(len(self.uses))}, " + \
-               f"op_name={repr(self.op.name)}, " + \
-               f"result_index={repr(self.result_index)}, name={repr(self.name)})"
+        return "{}<%{}: {}>(num_uses={}, op_name={})".format(
```
Why not f-strings?

math-fehr replied:
I would agree that on large strings like here, it is way more readable.

AntonLydike replied:
I actually dislike fstrings for most purposes. Purposefully splitting template and values improves readability enormously.

superlopuh reviewed (changes requested):
Could you please make the __repr__ as close as possible to the python expression you'd need to instantiate them? I don't see why we'd want to break the __repr__ contract

AntonLydike commented:
My rationale for the `%`:

I wanted to print the index for the block argument at first, and without any prefix it "looked wrong". I added the `%` prefix because I thought that it won't be confusing for people, since it's obvious that these won't stay the same and are highly dependent on the program you print.

I never wanted to print a "valid" number here, just the index, meaning either the index of the block argument, or the result index. I'll switch to a better format!

I changed that later, I'll push a newer version soon.

AntonLydike commented:
The new format with rationale:

## Block Argument:
```
<BlockArgument[0: !i32] with 0 uses>
```
 - Name, index, type of the argument
 - Then some human readable information
 - Everything in `<` brackets, so we stay within the contract of `repr` (meaining that `<` stuff is not 1:1 mappable)
 - Short, concise information for the user

## OpResult:
```
<OpResult[0: !i32] from arith.constant with 0 uses>
```

 - Very similar format to `BlockArgument`
 - Additionally prints op name where it's from

I'm not sure I 100% like the way the index is not prefixed by anything, but that's probably okay.
We currently don't print the name of the SSA Value, which might be okay.

georgebisbas commented on xdsl/ir.py:
```diff
@@ -204,9 +204,13 @@ def owner(self) -> Operation:
-        return f"OpResult(typ={repr(self.typ)}, num_uses={repr(len(self.uses))}, " + \
-               f"op_name={repr(self.op.name)}, " + \
-               f"result_index={repr(self.result_index)}, name={repr(self.name)})"
+        return "<{}[{}: {}] from {} with {} uses>".format(
```
I would add a comment here to explain why the printing format.

webmiche commented:
> ```
> <BlockArgument[0: !i32] with 0 uses>
> ```


In this example, what is what? Blockargument is the name, 0 is the index and !i32 is the type of the Blockargument? If so, I don't like it in the context of xDSL since `0: !i32` reads like and IntegerAttr to me. I would prefer something like this:

```
<Blockargument : !i32, index: 0, uses: 0>
```

Or even;


```
<Blockargument [type: !i32, index: 0, uses: 0]>
```

webmiche commented on xdsl/ir.py:
```diff
@@ -231,13 +235,12 @@ def owner(self) -> Block:
-                         f"num_blocks={len(self.block.ops)} ops)"
-        else:
-            block_repr = repr(self.block)
-        return f"OpResult(typ={repr(self.typ)}, num_uses={repr(len(self.uses))}" + \
```
IMO, to make this less controversial, I would first start with a small PR changin `OpResult` to `BlocKArgument` here. This is clearly a bug in the naming. Also the whole if-else thing doesn't make sene, IMO.

superlopuh reviewed (approved):
Thank you!
